File: graphics/particle_system.js
class ParticleSystem {
  constructor() {
    this.particles = [];
  }

  // Create a burst of particles at a given position
  createParticleBurst(position, numParticles) {
    for (let i = 0; i < numParticles; i++) {
      const particle = new Particle(position);
      this.particles.push(particle);
    }
  }

  // Update all particles in the system
  update() {
    for (let i = this.particles.length - 1; i >= 0; i--) {
      const particle = this.particles[i];
      particle.update();
      if (particle.isDead()) {
        this.particles.splice(i, 1);
      }
    }
  }

  // Render all particles in the system
  render() {
    for (const particle of this.particles) {
      particle.render();
    }
  }
}

class Particle {
  constructor(position) {
    this.position = position;
    this.velocity = createRandomVector(); // Function to create a random vector
    this.size = random(5, 15);
    this.color = randomColor(); // Function to generate a random color
    this.lifespan = 255; // Initial lifespan of the particle
  }

  update() {
    this.position.add(this.velocity);
    this.lifespan -= 2;
  }

  render() {
    // Render the particle as a colored circle
    fill(this.color.r, this.color.g, this.color.b, this.lifespan);
    noStroke();
    ellipse(this.position.x, this.position.y, this.size);
  }

  isDead() {
    return this.lifespan <= 0;
  }
}

// Helper function to create a random vector
function createRandomVector() {
  const angle = random(TWO_PI);
  const magnitude = random(1);
  return createVector(cos(angle) * magnitude, sin(angle) * magnitude);
}

// Helper function to generate a random color
function randomColor() {
  return {
    r: random(255),
    g: random(255),
    b: random(255),
  };
}

File: graphics/particle_system.cpp
#include <iostream>
#include <vector>

class Particle {
public:
    float x, y, z;
    float velocityX, velocityY, velocityZ;
    float size;
    float lifespan;
    // Другие параметры частицы

    void update(float deltaTime) {
        // Логика обновления частицы
        x += velocityX * deltaTime;
        y += velocityY * deltaTime;
        z += velocityZ * deltaTime;
        lifespan -= deltaTime;
        // Другие обновления
    }

    bool isAlive() const {
        return lifespan > 0.0f;
    }
};

class ParticleSystem {
public:
    std::vector<Particle> particles;
    // Другие параметры системы частиц

    void emitParticle() {
        Particle particle;
        // Инициализация параметров частицы
        particles.push_back(particle);
    }

    void update(float deltaTime) {
        for (int i = 0; i < particles.size(); ++i) {
            particles[i].update(deltaTime);
            if (!particles[i].isAlive()) {
                particles.erase(particles.begin() + i);
                --i;
            }
        }
        // Другие обновления системы
    }
};

int main() {
    ParticleSystem particleSystem;
    // Инициализация системы частиц и другие настройки

    while (true) {
        // Главный цикл игры
        float deltaTime = 0.016f; // Примерное значение времени шага
        particleSystem.emitParticle();
        particleSystem.update(deltaTime);
        // Отрисовка частиц и другие действия
    }

    return 0;
}
